packages/luna/src/components/Lessons/BookLesson/BookLessonDialog.tsx

```diff
@@ -156,80 +163,40 @@ export const BookLessonDialog: React.FC<{
     return { min, max };
   }, []);
 
-  const unpackedRules = useMemo(() => {
-    return unpackRules({
-      rules,
-      slots,
-      /**
-       * ! update {@link unpackRules} logic and make `start` date inclusive.
-       */
-      start: dateBounds.min.subtract(1, "day").toISOString(),
-      end: dateBounds.max.add(1, "day").toISOString(),
-    });
-  }, [dateBounds.max, dateBounds.min, rules, slots]);
+  const unBookedSlots = useMemo(
+    () => subtractSlots({ slots, subslots: bookedSlots }),
+    [slots, bookedSlots]
+  );
 
   const selectDaySlots = useCallback(
     (day: Dayjs) => {
-      const daySlots = unpackedRules.filter(
+      const daySlots = unBookedSlots.filter(
         (event) =>
           day.isSame(event.start, "day") || day.isSame(event.end, "day")
       );
 
-      const unpackedSlots: AttributedSlot[] = flattenDeep(
-        daySlots.map((rule) => splitRuleEvent(rule, duration))
-      ).map((event) => ({
-        ruleId: event.id,
-        slotId: event.id,
-        start: event.start,
-        end: event.end,
-        bookable: dayjs(event.start).isAfter(
-          dayjs().add(notice || 0, "minutes")
-        ),
-      }));
-
-      return unpackedSlots;
+      return getSubSlots(daySlots, duration);
     },
-    [duration, notice, unpackedRules]
+    [duration, unBookedSlots]
   );
 
   /**
-   * Date is considered valid incase it has at least one bookable (free) slot.
+   * Date is considered valid in case it has at least one bookable (free) slot.
    */
   const isValidDate = useCallback(
-    (date: Dayjs) =>
-      !isEmpty(selectDaySlots(date).filter((slot) => slot.bookable)),
+    (date: Dayjs) => !isEmpty(selectDaySlots(date)),
     [selectDaySlots]
   );
 
   /**
    * List of all slots including booked and not-yet-booked slots.
    */
   const allSlots = useMemo(() => {
-    const daySlots = selectDaySlots(date);
-    const bookedSlots: AttributedSlot[] = slots
-      .map((slot) => ({
-        ruleId: slot.ruleId,
-        slotId: slot.ruleId,
-        start: slot.start,
-        end: dayjs(slot.start).add(slot.duration, "minutes").toISOString(),
-        bookable: false,
-      }))
-      .filter(
-        (slot) => date.isSame(slot.start, "day") || date.isSame(slot.end, "day")
-      );
-    return Schedule.order(concat(daySlots, bookedSlots), "asc");
-  }, [selectDaySlots, date, slots]);
+    const availableSlots = selectDaySlots(date);
+    return orderSlots(availableSlots, "asc");
```

Comment:
I know that `orderSlots` returns `IAvailabilitySlot.GeneralSlot[]` but this misleading. Please update `orderSlots` to be generic. 

```ts
function orderSlots<T extends IAvailabilitySlot.Slot | IAvailabilitySlot.SubSlot>(slots: T[], dir: '...'): T[]
```